TST Extend tests for `scipy.sparse.*array` in `sklearn/cluster/tests/test_dbscan.py` + `test_birch.py` + `test_column_transformer.py`

msgomez06 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Towards [27090](https://github.com/scikit-learn/scikit-learn/issues/27090).

#### What does this implement/fix? Explain your changes.
This PR introduces sparse containers' list conditionally to the version of SciPy so that we can extend tests as part of [27090](https://github.com/scikit-learn/scikit-learn/issues/27090).

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

msgomez06 committed:
changed scipy.sparse to utils.fixes containers

msgomez06 committed:
Merge branch 'main' into skl_sprint

jjerphan commented on sklearn/cluster/tests/test_dbscan.py:
```diff
@@ -65,9 +65,9 @@ def test_dbscan_feature():
-
-def test_dbscan_sparse():
-    core_sparse, labels_sparse = dbscan(sparse.lil_matrix(X), eps=0.8, min_samples=10)
+pytest.mark.parametrize("lil_container", LIL_CONTAINERS)
```
```suggestion
@pytest.mark.parametrize("lil_container", LIL_CONTAINERS)
```

jjerphan reviewed (approved):
LGTM modulo a small fix and a green CI.

Thank you, @msgomez06. :slightly_smiling_face:

msgomez06 committed:
scipy.sparse to sklearn.utils.fixes in test_birch

ogrisel commented on sklearn/cluster/tests/test_dbscan.py:
```diff
@@ -108,10 +108,11 @@ def test_dbscan_sparse_precomputed_different_eps():
 @pytest.mark.parametrize("metric", ["precomputed", "minkowski"])
-def test_dbscan_input_not_modified(use_sparse, metric):
+@pytest.mark.parametrize("csr_container", CSR_CONTAINERS)
+def test_dbscan_input_not_modified(use_sparse, metric, csr_container):
     # test that the input is not modified by dbscan
     X = np.random.RandomState(0).rand(10, 10)
-    X = sparse.csr_matrix(X) if use_sparse else X
+    X = csr_container(X) if use_sparse else X
```
This could be optimized by not generating the cases 2 meaningless cases for `use_sparse=False, csr_container=csr_matrix` and `use_sparse=False, csr_container=csr_array`.

See #27099 for a similar example to drop `use_sparse` and instead parametrize `csr_container` with `CSR_CONTAINERS + [None]`.

ogrisel reviewed (commented):
Beyond Julien's suggestion we can also avoid introducing redundant parameter combinations as follows:

msgomez06 committed:
Merge branch 'scikit-learn:main' into skl_sprint

msgomez06 committed:
ran black 23.3.0, removed use_sparse in dbscan

msgomez06 commented:
regarding `test_column_transformer`:
I added the csr_container attribute to SparseMatrixTrans and added a pytest parameter when instantiating. However, I'm now running into a not-implemented error in `test_column_transformer_sparse_array` - running the test without it doesn't return other failures. 

> sklearn/compose/tests/test_column_transformer.py::test_column_transformer_sparse_array[csr_array] - NotImplementedError: We have not yet implemented 1D sparse slices; please index using explicit indices, e.g. `x[:, [0]]`

Thoughts on how to proceed?

msgomez06 committed:
made csr container attribute of SparseMatrixTrans

msgomez06 committed:
1D slicing not implemented workaround for test_column_transformer

msgomez06 committed:
Merge branch 'main' into skl_sprint

msgomez06 committed:
Merge branch 'main' into skl_sprint

ogrisel reviewed (approved):
LGTM as well. However to make it easier to review and merge incrementally, please open one PR per submodule next time (unless they are actually dependent on a common change/fix in the code base).
